### PR TITLE
cmd/dbconn: added in the SSL/TLS mode check behavior to enforce secure default

### PR DESCRIPTION
**Pull request by ryanluu12345**
Previously, SSL/TLS mode disabled was allowed by default. However, as part of providing guardrails for customers to be more secure, the new behavior defaults to disallowing insecure connections.

There is also a flag that allows that behavior to be overrided, --allow-tls-mode-disable, that can be used in cases where secure access is not possible (i.e. if source database was not set up with SSL in mind).

Resolves: CC-28386
Release Note: MOLT tools now enforce secure connections to databases as a default. There is a flag, --allow-tls-mode-disable, that allows customers to override that behavior if secure access is not possible.

Testing:
sslmode=disable as default -> tells you it's insecure
```
(⎈|rancher-desktop:N/A)ryanluu@crlMBP-XFFK6Q9QL6MTk1 molt % go run . fetch --source "$SOURCE" --target "$TARGET" --table-filter 'employees' --bucket-path 's3://molt-test' --table-handling truncate-if-exists --non-interactive --logging DEBUG --log-file "stdout"
{"level":"info","time":"2024-05-17T11:27:54-07:00","message":"default compression to GZIP"}
Error: TLS mode is set as insecure. If a secure TLS connection is not possible, enable the --allow-tls-mode-disable flag to override this check.
TLS mode is set as insecure. If a secure TLS connection is not possible, enable the --allow-tls-mode-disable flag to override this check.
exit status 1
```

override -> works
```
(⎈|rancher-desktop:N/A)ryanluu@crlMBP-XFFK6Q9QL6MTk1 molt % go run . fetch --source "$SOURCE" --target "$TARGET" --table-filter 'employees' --bucket-path 's3://molt-test' --table-handling truncate-if-exists --non-interactive --logging DEBUG --log-file "stdout" --allow-tls-mode-disable
{"level":"info","time":"2024-05-17T11:28:55-07:00","message":"default compression to GZIP"}
{"level":"debug","type":"data","flush_size":268435456,"flush_num_rows":0,"store":"*datablobstorage.s3Store","time":"2024-05-17T11:28:55-07:00","message":"initial config"}
{"level":"info","time":"2024-05-17T11:28:55-07:00","message":"checking database details"}
```

TODO:
- [x] Fix e2e tests
- [x] Rebase on the redaction PR
- [x] Update the integrated logic so that if we see an SSL mode disable error, we show that error instead of the generic connection error

### CODE REVIEW

File: dbconn/dbconn.go [hunk 1]
```diff
@@ -20,6 +20,12 @@ type ID string
 
 type OrderedConns [2]Conn
 
+type TLSModeDisableError struct{}
+
+func (s TLSModeDisableError) Error() string {
```
Comment (ryanluu12345) [inline dbconn/dbconn.go:25]: @taroface wondering if I could have your take on the copy here? I mainly want to let the user know that the TLS mode is not alllowed to be insecure by default. But if they must absolutely use insecure, we want them to specify this via a flag.

File: e2e/testdata/mysql/ssl_mode_default_error.ddt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+fetch --source 'mysql://user:password@0.0.0.0:3306/defaultdb' --target 'postgres://root@localhost:26257/defaultdb?sslmode=disable' --table-filter 'tbl1' --local-path /tmp/basic --local-path-listen-addr '0.0.0.0:9005' --non-interactive
+----
+{"level":"info","message":"default compression to GZIP"}
+Error: TLS mode is set to an insecure default. If you are unable to connect via TLS, use the --allow-tls-mode-disable flag to override this check.
+TLS mode is set to an insecure default. If you are unable to connect via TLS, use the --allow-tls-mode-disable flag to override this check.
+exit status 1
+
+# Explicitly specified ssl mode is disabled.
+verify --source 'mysql://user:password@0.0.0.0:3306/defaultdb?sslmode=disable' --target 'postgres://root@localhost:26257/defaultdb?sslmode=disable' --table-filter 'tbl1'
+----
```
Comment (ryanluu12345) [inline e2e/testdata/mysql/ssl_mode_default_error.ddt:11]: @ZhouXing19 any idea why the e2e test generates a double `----` line for this?

File: e2e/testdata/pg/ssl_mode_default_error.ddt [hunk 3]
```diff
@@ -0,0 +1,47 @@
+fetch --source 'postgres://postgres@localhost:5432/defaultdb' --target 'postgres://root@localhost:26257/defaultdb?sslmode=disable' --table-filter 'tbl1' --local-path /tmp/basic --local-path-listen-addr '0.0.0.0:9005' --non-interactive
+----
+{"level":"info","message":"default compression to GZIP"}
+Error: TLS mode is set to an insecure default. If you are unable to connect via TLS, use the --allow-tls-mode-disable flag to override this check.
+TLS mode is set to an insecure default. If you are unable to connect via TLS, use the --allow-tls-mode-disable flag to override this check.
+exit status 1
+
+verify --source 'postgres://postgres@localhost:5432/defaultdb' --target 'postgres://root@localhost:26257/defaultdb?sslmode=disable' --table-filter 'tbl1'
+----
```
Comment (ryanluu12345) [inline e2e/testdata/pg/ssl_mode_default_error.ddt:9]: Same question about the double dividers here too @ZhouXing19

File: e2e/testdata/mysql/ssl_mode_default_error.ddt (hunk 2 above)
Comment (ZhouXing19) [reply e2e/testdata/mysql/ssl_mode_default_error.ddt:11]: It's from the datadriven test when your result is blank iiuc:

```
		r.emit("----")
		if hasBlankLine(actual) {
			r.emit("----")
			r.rewrite.WriteString(actual)
			r.emit("----")
			r.emit("----")
			r.emit("")
		} else {
			// Here actual already ends in \n so emit adds a blank line.
			r.emit(actual)
		}
```

See also: https://github.com/cockroachdb/datadriven/blob/e384cf455877a2a2d83fd835940d4061a8d203b4/datadriven.go#L77-L89

Comment (ryanluu12345) [reply e2e/testdata/mysql/ssl_mode_default_error.ddt:11]: Ok, so it sounds expected. Any fixes I need to make here?

Comment (ZhouXing19) [reply e2e/testdata/mysql/ssl_mode_default_error.ddt:11]: I think trimming the blank line from the result should resolve it.

File: dbconn/pg.go [hunk 4]
```diff
@@ -26,14 +26,31 @@ type PGConn struct {
...
 	}
+
+	if err := checkIfPostgreTLSModeInsecure(cfg, allowTLSDisable); err != nil {
+		return nil, err
+	}
+
 	return ConnectPGConfig(ctx, id, cfg, false /* testOnly */)
 }
 
+func checkIfPostgreTLSModeInsecure(cfg *pgx.ConnConfig, tlsModeOveride bool) error {
```
Comment (Jeremyyang920) [inline dbconn/pg.go:42]:
```suggestion
func checkIfPostgresTLSModeInsecure(cfg *pgx.ConnConfig, tlsModeOveride bool) error {
```

File: dbconn/dbconn.go [hunk 5]
```diff
@@ -20,6 +20,12 @@ type ID string
...
+	return "TLS mode is set to an insecure default. If you are unable to connect via TLS, use the --allow-tls-mode-disable flag to override this check."
```
Comment (taroface) [inline dbconn/dbconn.go:26]:
```suggestion
	return "TLS mode is set as insecure. If a secure TLS connection is not possible, enable the --allow-tls-mode-disable flag to override this check."
```

**Review by taroface — COMMENTED** [review]
1 suggestion

File: dbconn/dbconn.go (hunk 5 above)
Comment (ryanluu12345) [reply dbconn/dbconn.go:26]: Updated!

File: cmd/internal/cmdutil/dbconn.go [hunk 6]
```diff
@@ -30,14 +31,28 @@ func RegisterDBConnFlags(cmd *cobra.Command) {
...
 	}
-	target, err := dbconn.Connect(ctx, "target", DBConnConfig.Target)
+	target, err := dbconn.Connect(ctx, "target", DBConnConfig.Target, allowTLSModeDisabled)
 	if err != nil {
 		return dbconn.OrderedConns{}, err
 	}
 	return dbconn.OrderedConns{source, target}, nil
 }
+
+func RegisterTLSModeFlags(cmd *cobra.Command) {
```
Comment (ryanluu12345) [inline cmd/internal/cmdutil/dbconn.go:46]: @taroface any recs here?

File: cmd/internal/cmdutil/dbconn.go [hunk 7]
```diff
@@ -30,14 +31,28 @@ func RegisterDBConnFlags(cmd *cobra.Command) {
...
+	cmd.PersistentFlags().BoolVar(
+		&allowTLSModeDisable,
+		"allow-tls-mode-disable",
+		false,
+		`Determines whether disabled SSL/TLS mode is allowed for connection to databases. Secure connections should be used by default.
```
Comment (taroface) [inline cmd/internal/cmdutil/dbconn.go:51]:
```suggestion
		`Allows insecure connection to databases. Secure SSL/TLS connections should be used by default. 
```

File: cmd/internal/cmdutil/dbconn.go [hunk 8]
```diff
@@ -30,14 +31,28 @@ func RegisterDBConnFlags(cmd *cobra.Command) {
...
 	return dbconn.OrderedConns{source, target}, nil
 }
+
+func RegisterTLSModeFlags(cmd *cobra.Command) {
+	cmd.PersistentFlags().BoolVar(
+		&allowTLSModeDisable,
+		"allow-tls-mode-disable",
+		false,
+		`Determines whether disabled SSL/TLS mode is allowed for connection to databases. Secure connections should be used by default. 
+Provide this flag if either the source and target cannot be accessed via SSL/TLS`,
```
Comment (taroface) [inline cmd/internal/cmdutil/dbconn.go:52]:
```suggestion
Enable this flag only if you cannot access the source or target database via secure SSL/TLS`,
```

File: e2e/testdata/mysql/ssl_mode_default_error.ddt (hunk 2 above)
Comment (ryanluu12345) [reply e2e/testdata/mysql/ssl_mode_default_error.ddt:11]: Ok, I'll try that.

Comment (ryanluu12345) [reply e2e/testdata/mysql/ssl_mode_default_error.ddt:11]: I got rid of the extra new lines but still see it:
image

Comment (ryanluu12345) [reply e2e/testdata/mysql/ssl_mode_default_error.ddt:11]: When I remove all new lines in general, it wipes the output for `verify`. It's a bit weird. I think I'll just implement it as is. We can look into this later.

File: cmd/internal/cmdutil/dbconn.go (hunk 8 above)
Comment (ryanluu12345) [reply cmd/internal/cmdutil/dbconn.go:52]: Noted!

**Review by taroface — APPROVED** [review]
copy lgtm

File: cmd/internal/cmdutil/dbconn.go (hunk 7 above)
Comment (ryanluu12345) [reply cmd/internal/cmdutil/dbconn.go]: Done

**Review by rsevinsky-cr — APPROVED** [review]

**Pull request merged**